### PR TITLE
Added AlphaImage support

### PR DESCRIPTION
**Pull request by Mailaender**
Initial support for http://modenc.renegadeprojects.com/AlphaImage which is basically a grey scale palette and a configurable background index (middle not start is neutral/transparent here) yet.

The result does not look too impressive yet, but I guess it is a start:

image

### CODE REVIEW

**Comment by Mailaender** [comment]
Updated with the help of @tomsons26:

image

**Comment by x-a-n-a-x** [comment]
Gorgeous.

**Comment by reaperrr** [comment]
I think it would look better if the alpha image was centered at the base of the light post rather than the top, otherwise this looks great.

File: OpenRA.Mods.Common/Traits/World/PaletteFromFile.cs [hunk 1]
```diff
@@ -18,12 +18,22 @@ class PaletteFromFileInfo : ITraitInfo
...
 		[Desc("filename to load")]
 		public readonly string Filename = null;
+
 		[Desc("Map listed indices to shadow. Ignores previous color.")]
 		public readonly int[] ShadowIndex = { };
+
+		[Desc("Converted to transparency.")]
+		public readonly int BackgroundIndex = 0;
+
+		[Desc("Converted to transparency.")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/World/PaletteFromFile.cs]: This isn't true.
Perhaps something like 'Converts `BackgroundIndex` to transparency.` would be better.

File: OpenRA.Game/Graphics/Palette.cs [hunk 2]
```diff
@@ -97,18 +97,18 @@ public void CopyToArray(Array destination, int destinationOffset)
 			Buffer.BlockCopy(colors, 0, destination, destinationOffset * 4, Palette.Size * 4);
 		}
 
-		public ImmutablePalette(string filename, int[] remap)
+		public ImmutablePalette(string filename, int[] remap, int background = 0, uint color = 0)
```
Comment (pchote) [inline OpenRA.Game/Graphics/Palette.cs]: It seems rather silly to add this extra plumbing in order to support a palette loaded from a custom file.  Just build the custom palette in code instead - like we do for our other special palette types.

**Comment by pchote** [comment]
I suggest squashing these commits down to one that adds the new blend modes, and another that adds the new palette and light post.

File: OpenRA.Game/Graphics/Palette.cs (hunk 2 above)
Comment (Mailaender) [inline OpenRA.Game/Graphics/Palette.cs]: I don't need this anymore with the fixed palette so I will just remove it. Saving to a local branch `configurable-palette-background` for later usage.

**Comment by Mailaender** [comment]
Updated.

**Comment by Phrohdoh** [comment]
screen shot 2015-04-04 at 1 36 16 pm

**Comment by Mailaender** [comment]
See http://imgur.com/a/bn6BB for more stuff you can do with this.

**Comment by pchote** [comment]
:+1: code and effect looks good.  This was one of the last prerequisites before doing TS-style per-cell lighting, so we should be able to get that done during the next release cycle.

**Comment by Phrohdoh** [comment]
Is there a fix to the artifact I posted?

**Comment by reaperrr** [comment]
I took TS' ```alphatst.shp``` and changed the canvas from 278x139 to 278x140 in SHP Builder and that black line was gone, changed it back to 139 and it reappeared. Looks like something has trouble with odd numbers, I've occasionally (but not often) spotted similar lines around voxels, too (yes, vertical ones as well). Don't know whether there's a problem with OpenRA's renderer or the shp(ts) loader (though I'd suspect the former due to the voxel issue I mentioned), but I'd say this isn't an issue created by this PR.

I've pushed and pr'd the fixed ```alphatst.shp``` here: https://github.com/Mailaender/OpenRA/pull/54

**Comment by Phrohdoh** [comment]
:+1: once Mailaender#54 is merged into this.

**Comment by pchote** [comment]
At a risk of reigniting an old argument, adding alphatst.shp makes this lose my :+1:.  It's clear that this is caused by (I bet a simple) bug in our engine, and i'm completely against hiding this behind a hacked sprite file.

**Comment by Mailaender** [comment]
According to http://modenc.renegadeprojects.com/AlphaImage#Using_Alpha_Images the game files don't guarantee to contain a correctly named `alphatst.shp` so it may be better if we ship our own. However I also would like to investigate potential SHP image reading offset issues. Not sure if this is the right pull request and an ideal test case though.

**Comment by reaperrr** [comment]
Opened a ticket for that black line issue. #7883

**Comment by Mailaender** [comment]
Rebased.

**Comment by LipkeGu** [comment]
is it able to "smooth out" the alpha image on the borders? :) so that there is no light border? as like in vanilla TS?

**Comment by Mailaender** [comment]
You would have to modify the SHP for that.

**Comment by Mailaender** [comment]
Just to be clear you can do lot's of things with this with the correct template:

Imgur

or vehicle lights

Imgur

or red zone bad weather clouds

Imgur

**Comment by x-a-n-a-x** [comment]
And maybe some people will want to use this effect on rotors:
orcafighter

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/f8f5711b6c5bc8c4f7fabebd436c4f5c446d16ce%5E...f8f5711b6c5bc8c4f7fabebd436c4f5c446d16ce

**Pull request merged**